tools/indexscan/main.go

```diff
@@ -109,6 +110,81 @@ type searchStrategy interface {
 	Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error)
 }
 
+// DynamicTreeSearchStrategy  searches all TreeObjects in the repository for
+// the number of matches against the input files provided.
+type DynamicTreeSearchStrategy struct {
+}
+
+// Search returns the set of matching commits along with the number of matches are returned.
+func (DynamicTreeSearchStrategy) Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error) {
+	files := make(map[plumbing.Hash]bool)
+	for _, h := range hashes {
+		_, err = r.BlobObject(plumbing.NewHash(h))
+		if err == plumbing.ErrObjectNotFound {
+			// Skip files not present in repo.
+			err = nil
+		} else if err != nil {
+			return
+		} else {
+			files[plumbing.NewHash(h)] = true
+		}
+	}
+	total = len(files)
+	if total == 0 {
+		err = errors.New("repo contains no matching files")
+		return
+	}
+	// Construct cache of all trees and their match count.
+	cache := make(map[plumbing.Hash]int)
+	ti, _ := r.TreeObjects()
+	ti.ForEach(func(t *object.Tree) error {
+		countTree(t, files, cache)
+		return nil
+	})
+	// Search through all commits for tree with closest match.
```

Comment:
Search through all commits *in the* tree *for the* closest match?

When I first read this and the function signature I thought it was looking for a tree but I think it's looking for a commit.

Reply:
no a commit has a single tree associated with it. this phrasing implies commits and trees are interchangable which, considering their 1:1 correspondence, doesn't feel like too much of a leap. i'm fine with either phrasing.